In [ ]:
import pandas as pd
import numpy as np
import matplotlib as plt

#import math as m
from math import *
In the first manner, we have defined an alias m to library math. We can now use various functions from math library (e.g. factorial) by referencing it using the alias m.factorial().

In the second manner, you have imported the entire name space in math i.e. you can directly use factorial() without referring to math.

In [3]:
df = pd.read_csv("Downloads//train_.csv")

In [4]:
df.head(10)

Loan_ID Gender Married Dependents     Education Self_Employed  \
0  LP001002   Male      No          0      Graduate            No   
1  LP001003   Male     Yes          1      Graduate            No   
2  LP001005   Male     Yes          0      Graduate           Yes   
3  LP001006   Male     Yes          0  Not Graduate            No   
4  LP001008   Male      No          0      Graduate            No   
5  LP001011   Male     Yes          2      Graduate           Yes   
6  LP001013   Male     Yes          0  Not Graduate            No   
7  LP001014   Male     Yes         3+      Graduate            No   
8  LP001018   Male     Yes          2      Graduate            No   
9  LP001020   Male     Yes          1      Graduate            No   

   ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
0             5849                0.0         NaN             360.0   
1             4583             1508.0       128.0             360.0   
2             3000                0.0        66.0             360.0   
3             2583             2358.0       120.0             360.0   
4             6000                0.0       141.0             360.0   
5             5417             4196.0       267.0             360.0   
6             2333             1516.0        95.0             360.0   
7             3036             2504.0       158.0             360.0   
8             4006             1526.0       168.0             360.0   
9            12841            10968.0       349.0             360.0   

   Credit_History Property_Area Loan_Status  
0             1.0         Urban           Y  
1             1.0         Rural           N  
2             1.0         Urban           Y  
3             1.0         Urban           Y  
4             1.0         Urban           Y  
5             1.0         Urban           Y  
6             1.0         Urban           Y  
7             0.0     Semiurban           N  
8             1.0         Urban           Y  
9             1.0     Semiurban           N

In [5]:
df.describe()

ApplicantIncome  CoapplicantIncome  LoanAmount  Loan_Amount_Term  \
count       614.000000         614.000000  592.000000         600.00000   
mean       5403.459283        1621.245798  146.412162         342.00000   
std        6109.041673        2926.248369   85.587325          65.12041   
min         150.000000           0.000000    9.000000          12.00000   
25%        2877.500000           0.000000  100.000000         360.00000   
50%        3812.500000        1188.500000  128.000000         360.00000   
75%        5795.000000        2297.250000  168.000000         360.00000   
max       81000.000000       41667.000000  700.000000         480.00000   

       Credit_History  
count      564.000000  
mean         0.842199  
std          0.364878  
min          0.000000  
25%          1.000000  
50%          1.000000  
75%          1.000000  
max          1.000000

#describe() function would provide count, mean, standard deviation (std), min, quartiles and max in its output, can infer LoanAmount has (614 – 592) 22 missing values.We can also look that about 84% applicants have a credit_history. How? The mean of Credit_History field is 0.84 (Credit_History has value 1 for those who have a credit history and 0 otherwise)

In [6]:
df['Property_Area'].value_counts()

Semiurban    233
Urban        202
Rural        179
Name: Property_Area, dtype: int64

#prints frequency table
## dfname[‘column_name’] is a basic indexing technique to acess a particular column of the dataframe.

In [8]:
#plot histogrmam of applicant income
%matplotlib
df['ApplicantIncome'].hist(bins=50)

Using matplotlib backend: MacOSX


In [9]:
#plot boxplot to look at distributions, in this case split by eduction
df.boxplot(column='ApplicantIncome', by = 'Education')

## Catagorical Variable analysis
#May need to come back to this later - (https://www.analyticsvidhya.com/blog/2016/01/complete-tutorial-learn-data-science-python-scratch-2/) 
Excel style pivot tables 
Example below looks at the chance of getting a loan based on credit history (which can be 0 or 1)

In [11]:
temp1 = df['Credit_History'].value_counts(ascending=True)
temp2 = df.pivot_table(values='Loan_Status',index=['Credit_History'],aggfunc=lambda x: x.map({'Y':1,'N':0}).mean())
print ('Frequency Table for Credit History:') 
print (temp1)

print ('\nProbility of getting loan for each Credit History class:') 
print (temp2)

Frequency Table for Credit History:
0.0     89
1.0    475
Name: Credit_History, dtype: int64

Probility of getting loan for each Credit History class:
                Loan_Status
Credit_History             
0.0                0.078652
1.0                0.795789


In [12]:
import matplotlib.pyplot as plt
fig = plt.figure(figsize=(8,4))
ax1 = fig.add_subplot(121)
ax1.set_xlabel('Credit_History')
ax1.set_ylabel('Count of Applicants')
ax1.set_title("Applicants by Credit_History")
temp1.plot(kind='bar')

ax2 = fig.add_subplot(122)
temp2.plot(kind = 'bar')
ax2.set_xlabel('Credit_History')
ax2.set_ylabel('Probability of getting loan')
ax2.set_title("Probability of getting loan by credit history")


## Data Munging

In [13]:
#checks the number of nulls/NaNs in the dataset
df.apply(lambda x: sum(x.isnull()),axis=0) 

Loan_ID               0
Gender               13
Married               3
Dependents           15
Education             0
Self_Employed        32
ApplicantIncome       0
CoapplicantIncome     0
LoanAmount           22
Loan_Amount_Term     14
Credit_History       50
Property_Area         0
Loan_Status           0
dtype: int64

#for more on how to treat missing values see https://www.analyticsvidhya.com/blog/2016/01/complete-tutorial-learn-data-science-python-scratch-2/ and more detail https://www.analyticsvidhya.com/blog/2016/01/guide-data-exploration/

## Predictive Models
#For brief into to model types and setup see bottom of page https://www.analyticsvidhya.com/blog/2016/01/complete-tutorial-learn-data-science-python-scratch-2/

In [18]:
#Boolean Indexing - want to filter values of a column based on conditions from another set of columns
df.loc[(df["Gender"]=="Female") & (df["Education"]=="Not Graduate") & (df["Loan_Status"]=="Y"), ["Gender","Education","Loan_Status"]]

Gender     Education Loan_Status
50   Female  Not Graduate           Y
197  Female  Not Graduate           Y
205  Female  Not Graduate           Y
279  Female  Not Graduate           Y
403  Female  Not Graduate           Y
407  Female  Not Graduate           Y
439  Female  Not Graduate           Y
463  Female  Not Graduate           Y
468  Female  Not Graduate           Y
480  Female  Not Graduate           Y
493  Female  Not Graduate           Y
534  Female  Not Graduate           Y
544  Female  Not Graduate           Y
587  Female  Not Graduate           Y

## Merge data frames

In [19]:
# in this case information being added from a different source on average property rates. First define dataframe:
prop_rates = pd.DataFrame([1000, 5000, 12000], index=['Rural','Semiurban','Urban'],columns=['rates'])
prop_rates

rates
Rural       1000
Semiurban   5000
Urban      12000

In [21]:
#now merge infor into original data frame
df_merged = df.merge(right=prop_rates, how='inner',left_on='Property_Area',right_index=True, sort=False)
df_merged.pivot_table(values='Credit_History',index=['Property_Area','rates'], aggfunc=len)

Credit_History
Property_Area rates                
Rural         1000            179.0
Semiurban     5000            233.0
Urban         12000           202.0

# Second example
#https://pandas.pydata.org/pandas-docs/stable/10min.html

In [23]:
import matplotlib.pyplot as plt

In [24]:
#create a series from a list
s = pd.Series([1,3,5,np.nan,6,8])
s

0    1.0
1    3.0
2    5.0
3    NaN
4    6.0
5    8.0
dtype: float64

In [26]:
#create dataframe  by passing a numpy aray with datetime index and labled columns
dates = pd.date_range('20130101', periods=6)
dates

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06'],
              dtype='datetime64[ns]', freq='D')

In [28]:
newdf = pd.DataFrame(np.random.randn(6,4), index=dates, columns=list('ABCD'))
newdf

A         B         C         D
2013-01-01 -0.663034  0.780399  0.271732  1.183859
2013-01-02  0.105860 -0.832518  1.054591 -1.855821
2013-01-03  0.221425  1.164069 -2.181819 -1.308305
2013-01-04  0.884620 -0.115030 -0.295956 -0.681401
2013-01-05  0.415191 -0.230046 -1.097490  0.157465
2013-01-06  0.632692 -1.685889 -0.249395 -0.095298

In [34]:
newdf2 = pd.DataFrame(np.random.randn(4,4), index = ('apple', 'orange','kiwi','banana'), columns = list('1234'))
newdf2

1         2         3         4
apple  -0.723615  0.266046 -0.375645  0.752161
orange  0.038535  1.223214  0.189847 -0.922385
kiwi   -0.815646 -1.152559  2.041845  0.168393
banana  0.733454 -0.534694 -0.425132 -1.286101

In [35]:
newdf2.dtypes

1    float64
2    float64
3    float64
4    float64
dtype: object

In [37]:
newdf.tail(3)

A         B         C         D
2013-01-04  0.884620 -0.115030 -0.295956 -0.681401
2013-01-05  0.415191 -0.230046 -1.097490  0.157465
2013-01-06  0.632692 -1.685889 -0.249395 -0.095298

In [38]:
#listing indexes and columns

newdf.columns

Index(['A', 'B', 'C', 'D'], dtype='object')

In [39]:
newdf.index

DatetimeIndex(['2013-01-01', '2013-01-02', '2013-01-03', '2013-01-04',
               '2013-01-05', '2013-01-06'],
              dtype='datetime64[ns]', freq='D')

In [40]:
newdf.values

array([[-0.66303447,  0.78039879,  0.27173213,  1.18385876],
       [ 0.10586043, -0.8325181 ,  1.05459124, -1.85582135],
       [ 0.22142544,  1.16406932, -2.1818187 , -1.30830549],
       [ 0.88462016, -0.1150304 , -0.29595578, -0.68140142],
       [ 0.41519062, -0.23004553, -1.0974903 ,  0.15746541],
       [ 0.63269186, -1.68588945, -0.24939511, -0.09529757]])

In [41]:
newdf.describe()

A         B         C         D
count  6.000000  6.000000  6.000000  6.000000
mean   0.266126 -0.153169 -0.416389 -0.433250
std    0.534817  1.041612  1.118957  1.088964
min   -0.663034 -1.685889 -2.181819 -1.855821
25%    0.134752 -0.681900 -0.897107 -1.151579
50%    0.318308 -0.172538 -0.272675 -0.388349
75%    0.578317  0.556541  0.141450  0.094275
max    0.884620  1.164069  1.054591  1.183859

In [42]:
#Transposing data
newdf.T

2013-01-01  2013-01-02  2013-01-03  2013-01-04  2013-01-05  2013-01-06
A   -0.663034    0.105860    0.221425    0.884620    0.415191    0.632692
B    0.780399   -0.832518    1.164069   -0.115030   -0.230046   -1.685889
C    0.271732    1.054591   -2.181819   -0.295956   -1.097490   -0.249395
D    1.183859   -1.855821   -1.308305   -0.681401    0.157465   -0.095298

In [47]:
#Sorting by an axis
newdf.sort_index(axis=1,ascending=False)

D         C         B         A
2013-01-01  1.183859  0.271732  0.780399 -0.663034
2013-01-02 -1.855821  1.054591 -0.832518  0.105860
2013-01-03 -1.308305 -2.181819  1.164069  0.221425
2013-01-04 -0.681401 -0.295956 -0.115030  0.884620
2013-01-05  0.157465 -1.097490 -0.230046  0.415191
2013-01-06 -0.095298 -0.249395 -1.685889  0.632692

In [49]:
newdf.sort_values(by='D')

A         B         C         D
2013-01-02  0.105860 -0.832518  1.054591 -1.855821
2013-01-03  0.221425  1.164069 -2.181819 -1.308305
2013-01-04  0.884620 -0.115030 -0.295956 -0.681401
2013-01-06  0.632692 -1.685889 -0.249395 -0.095298
2013-01-05  0.415191 -0.230046 -1.097490  0.157465
2013-01-01 -0.663034  0.780399  0.271732  1.183859

In [51]:
newdf['D']

2013-01-01    1.183859
2013-01-02   -1.855821
2013-01-03   -1.308305
2013-01-04   -0.681401
2013-01-05    0.157465
2013-01-06   -0.095298
Freq: D, Name: D, dtype: float64

In [52]:
#selecting by lable 
newdf.loc[dates[0]]

A   -0.663034
B    0.780399
C    0.271732
D    1.183859
Name: 2013-01-01 00:00:00, dtype: float64

In [54]:
#Selecting on a multi-axis by label
newdf.loc[:,['A','B']]

A         B
2013-01-01 -0.663034  0.780399
2013-01-02  0.105860 -0.832518
2013-01-03  0.221425  1.164069
2013-01-04  0.884620 -0.115030
2013-01-05  0.415191 -0.230046
2013-01-06  0.632692 -1.685889

In [56]:
newdf.loc['20130102':'20130104',['A','B']]

A         B
2013-01-02  0.105860 -0.832518
2013-01-03  0.221425  1.164069
2013-01-04  0.884620 -0.115030

In [61]:
#the equivalent selecting by location
newdf.iloc[1:4,0:2]

A         B
2013-01-02  0.105860 -0.832518
2013-01-03  0.221425  1.164069
2013-01-04  0.884620 -0.115030

In [63]:
newdf.iloc[1:3,:]

A         B         C         D
2013-01-02  0.105860 -0.832518  1.054591 -1.855821
2013-01-03  0.221425  1.164069 -2.181819 -1.308305

In [65]:
newdf.iloc[1,1]

-0.83251809521341336

In [69]:
#can use a single columns values to select data
newdf[newdf.A < 0]

A         B         C         D
2013-01-01 -0.663034  0.780399  0.271732  1.183859

In [70]:
newdf_add = newdf.copy()
newdf_add['E'] = ['one', 'one','two','three','four','three']
newdf_add

A         B         C         D      E
2013-01-01 -0.663034  0.780399  0.271732  1.183859    one
2013-01-02  0.105860 -0.832518  1.054591 -1.855821    one
2013-01-03  0.221425  1.164069 -2.181819 -1.308305    two
2013-01-04  0.884620 -0.115030 -0.295956 -0.681401  three
2013-01-05  0.415191 -0.230046 -1.097490  0.157465   four
2013-01-06  0.632692 -1.685889 -0.249395 -0.095298  three

In [73]:
#setting values by label
newdf.at[dates[0],'A'] = 0
newdf.head(3)

A         B         C         D
2013-01-01  0.000000  0.780399  0.271732  1.183859
2013-01-02  0.105860 -0.832518  1.054591 -1.855821
2013-01-03  0.221425  1.164069 -2.181819 -1.308305

In [75]:
#setting values by position
newdf.iat[0,1] = 0
newdf.head(3)

A         B         C         D
2013-01-01  0.000000  0.000000  0.000000  1.183859
2013-01-02  0.105860 -0.832518  1.054591 -1.855821
2013-01-03  0.221425  1.164069 -2.181819 -1.308305

In [77]:
#Reindexing allows you to change/add/delete the index on a specified axis. This returns a copy of the data.
df1 = newdf.reindex(index=dates[0:4], columns=list(newdf.columns) + ['E'])
df1.loc[dates[0]:dates[1],'E'] = 1
df1

A         B         C         D    E
2013-01-01  0.000000  0.000000  0.000000  1.183859  1.0
2013-01-02  0.105860 -0.832518  1.054591 -1.855821  1.0
2013-01-03  0.221425  1.164069 -2.181819 -1.308305  NaN
2013-01-04  0.884620 -0.115030 -0.295956 -0.681401  NaN

In [78]:
#drop any rows that have missing data
df1.dropna(how='any')

A         B         C         D    E
2013-01-01  0.00000  0.000000  0.000000  1.183859  1.0
2013-01-02  0.10586 -0.832518  1.054591 -1.855821  1.0

In [79]:
#Filling missing data
df1.fillna(value=5)

A         B         C         D    E
2013-01-01  0.000000  0.000000  0.000000  1.183859  1.0
2013-01-02  0.105860 -0.832518  1.054591 -1.855821  1.0
2013-01-03  0.221425  1.164069 -2.181819 -1.308305  5.0
2013-01-04  0.884620 -0.115030 -0.295956 -0.681401  5.0